File: Python/problem033.py
#! /usr/bin/env python3
'''
Problem 33 - Project Euler
http://projecteuler.net/index.php?section=problems&id=033
'''
from math import gcd
from fractions import Fraction
from functools import reduce
from operator  import mul

def is_digit_cancelling_fraction_pair(i, j):
    i_s = str(i)
    j_s = str(j)
    for s in i_s:
        if s in j_s and i_s[0] != i_s[1] and j_s[0] != j_s[1]:
            i2 = int(i_s.replace(s,''))
            j2 = int(j_s.replace(s,''))

            if i % 10 != 0 and j % 10 != 0 and Fraction(i,j) == Fraction(i2,j2):
                return True
    return False


if __name__ == '__main__':
    pairs = [(i, j) for i in range(10, 100) for j in range(i+1,100)]
    p2 = list(filter(lambda x: is_digit_cancelling_fraction_pair(*x), pairs))
    print(p2)
    print(reduce(mul, map(lambda x: Fraction(x[0], x[1]), p2)).denominator)

File: Clojure/problem033.clj
#! /usr/bin/env clojure
; Problem 33 - Project Euler
; http://projecteuler.net/index.php?section=problems&id=33
(require '[clojure.string :as s])

(defn digit-cancelling-fraction-pair? [nd]
  (letfn [(s2i [s] (if (> (count s) 0)
                     (Integer/parseInt s)
                     1))]
    (let [n (first nd)
          d (second nd)
          d0 (s2i (str (first (str d))))
          d1 (s2i (str (second (str d))))
          n_d0 (s2i (s/replace (str n) (str d0) ""))
          n_d1 (s2i (s/replace (str n) (str d1) ""))]
      (and (or (and (> d0 0) (= (/ n d) (/ n_d1 d0)))
               (and (> d1 0) (= (/ n d) (/ n_d0 d1))))
           (and (not (zero? (mod n 10)))
                (not (zero? (mod d 10))))))))

(->> (for [x (range 10 100)
           y (range (inc x) 100)]
       (list x y))
  (filter digit-cancelling-fraction-pair?)
  (map #(/ (first %) (second %)))
  (reduce *)
  (denominator)
  (println))
